drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_npcx.c

```diff
@@ -0,0 +1,998 @@
+	inst->EVSTAT = 0xff;
+
+	npcx_miwu_interrupt_configure(&config->shi_cs_wui, NPCX_MIWU_MODE_EDGE, NPCX_MIWU_TRIG_LOW);
+
+	/* SHI interrupt installation */
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), shi_npcx_isr, DEVICE_DT_INST_GET(0),
+		    0);
+
+	shi_npcx_enable(dev);
+
+	return ret;
+}
+
+static int shi_npcx_init_peripheral(const struct device *dev)
+{
+	struct shi_npcx_data *data = dev->data;
+
+	/* Allow writing to rx buff at startup and block on reading. */
+	k_sem_init(&data->handler_owns, 0, 1);
+	k_sem_init(&data->dev_owns, 1, 1);
+
+	return 0;
+}
+
+static int shi_npcx_init(const struct device *dev)
+{
+	int ret;
+
+	ret = shi_npcx_init_registers(dev);
+	if (ret) {
+		return ret;
+	}
+
+	ret = shi_npcx_init_peripheral(dev);
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int shi_npcx_periph_init_context(const struct device *dev,
+					struct ec_host_cmd_periph_rx_ctx *rx_ctx)
+{
+	struct shi_npcx_data *data = dev->data;
+
+	rx_ctx->dev_owns = &data->dev_owns;
+	rx_ctx->handler_owns = &data->handler_owns;
+	rx_ctx->buf = data->in_msg;
+	rx_ctx->len = &data->sz_received;
+
+	return 0;
+}
+
+static int shi_npcx_periph_send(const struct device *dev,
+				const struct ec_host_cmd_periph_tx_buf *in_buf)
+{
+	struct shi_npcx_data *data = dev->data;
+	uint8_t *out_buf = data->out_msg + EC_SPI_FRAME_START_LENGTH;
+
+	/*
+	 * Disable interrupts. This routine is not called from interrupt
+	 * context and buffer underrun will likely occur if it is
+	 * preempted after writing its initial reply byte. Also, we must be
+	 * sure our state doesn't unexpectedly change, in case we're expected
+	 * to take RESP_NOT_RDY actions.
+	 */
+	__disable_irq();
+
+	memcpy(out_buf, in_buf->buf, in_buf->len);
+
+	if (data->state == SHI_STATE_PROCESSING) {
+		/* Append our past-end byte, which we reserved space for. */
+		((uint8_t *)out_buf)[in_buf->len] = EC_SPI_PAST_END;
+
+		/* Computing sending bytes of response */
+		data->sz_response = in_buf->len + SHI_PROTO3_OVERHEAD;
+
+		/* Start to fill output buffer with msg buffer */
+		shi_write_first_pkg_outbuf(dev, data->sz_response);
+
+		/* Transmit the reply */
+		data->state = SHI_STATE_SENDING;
+		LOG_DBG("SND-");
+	} else if (data->state == SHI_STATE_CNL_RESP_NOT_RDY) {
+		/*
+		 * If we're not processing, then the AP has already terminated
+		 * the transaction, and won't be listening for a response.
+		 * Reset state machine for next transaction.
+		 */
+		shi_npcx_reset_prepare(dev);
+		LOG_DBG("END\n");
+	} else
+		LOG_ERR("Unexpected state %d in response handler", data->state);
+
+	__enable_irq();
+	return 0;
+}
+
+static const struct ec_host_cmd_periph_api ec_host_cmd_api = {
+	.init = shi_npcx_periph_init_context,
+	.send = shi_npcx_periph_send,
+};
+
+#ifdef CONFIG_PM_DEVICE
+static int shi_npcx_pm_cb(const struct device *dev, enum pm_device_action action)
+{
+	int ret = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		shi_npcx_disable(dev);
```

Comment:
I understand, these callbacks are for device power management, but they will be enabled/disabled based on the host power state. There's no other way to allow to disable and enable the interface outside the driver code.
Here's how it looks like in the EC code: https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/3827184

Reply:
Thank you for the explanation of using the PM_DEVICE* power management to disable/enable the SHI.
I understand now how it works with the HOST's power state.
Does it make sense to use  pm_device_runtime_disable(set SHI PM_STATE to RESUME)/pm_device_runtime_enable(set SHI PM_STATE to SUSPEND) instead of pm_device_runtime_get/pm_device_runtime_out to enable/disable SHI?
I asked this because I read the the [pm device mangement document](https://docs.zephyrproject.org/3.1.0/services/pm/device_runtime.html#introduction) and saw:
```
Therefore, applications can not manually suspend or resume a device. An application can, however, decide when to  
disable or enable runtime power management for a device. This can be useful, for example, if an application wants a
particular device to be always active.
```

Reply:
> Does it make sense to use pm_device_runtime_disable(set SHI PM_STATE to RESUME)/pm_device_runtime_enable(set SHI PM_STATE to SUSPEND) instead of pm_device_runtime_get/pm_device_runtime_out to enable/disable SHI?

Sure, it sounds like I shouldn't use the put/get in the EC application, even if it isn't used inside the HC driver. I will make a change to the code there. It shouldn't require any change in this PR.
Thank you very much for checking the code and mentioning this documentation, I wasn't aware of that.

Reply:
I think you need to add a call to `pm_device_init_suspended()` as part of `shi_npcx_init()` so that when the EC application calls `pm_device_runtime_disable()`, the PM layer will transition the SHI driver from suspended to active.   [PM device implementation guidelines](https://docs.zephyrproject.org/latest/services/pm/device_runtime.html#implementation-guidelines)

Reply:
Yea, right, otherwise it would fire the callback unnecessarily. But the enable and disable are supposed to only be in the EC application, right?

Reply:
I think the driver init is supposed to call both `pm_device_init_suspended() and `pm_device_runtime_enable()`.  This ensures the internal PM state is initialized correctly.

Example [here](https://github.com/zephyrproject-rtos/zephyr/blob/42db096d284016b9352965ddabe82148e5d614e2/drivers/dai/intel/dmic/dmic.c#L877).  

I find only [instance](https://github.com/zephyrproject-rtos/zephyr/blob/42db096d284016b9352965ddabe82148e5d614e2/drivers/power_domain/power_domain_gpio.c#L99) where `pm_device_runtime_enable()` isn't called - so I assume this is a bug.